# Dataset Information

Dataset: Annual CO₂ emissions (fossil CO₂)

Source: Global Carbon Project, "Global Carbon Budget v15", with major processing by Our World in Data.

License: CC BY 4.0

Access method: CSV file hosted by Our World in Data (OWID).

Link: https://ourworldindata.org/co2-and-greenhouse-gas-emissions

## Load dataset

In [2]:
import pandas as pd

#Fetch the data
url = "https://raw.githubusercontent.com/owid/co2-data/master/owid-co2-data.csv"
df = pd.read_csv(url)

## Saving raw dataset

In [3]:
df.to_csv(
    "../data/raw/owid_co2_data.csv",
    index=False
)

## The purpose of this section is to understand the dataset structure, verify data quality, inspect data types, identify missing values, and determine which variables are relevant for further analysis.

### Checking columns in the dataset. 

In [4]:
df.columns.tolist()

['country',
 'year',
 'iso_code',
 'population',
 'gdp',
 'cement_co2',
 'cement_co2_per_capita',
 'co2',
 'co2_growth_abs',
 'co2_growth_prct',
 'co2_including_luc',
 'co2_including_luc_growth_abs',
 'co2_including_luc_growth_prct',
 'co2_including_luc_per_capita',
 'co2_including_luc_per_gdp',
 'co2_including_luc_per_unit_energy',
 'co2_per_capita',
 'co2_per_gdp',
 'co2_per_unit_energy',
 'coal_co2',
 'coal_co2_per_capita',
 'consumption_co2',
 'consumption_co2_per_capita',
 'consumption_co2_per_gdp',
 'cumulative_cement_co2',
 'cumulative_co2',
 'cumulative_co2_including_luc',
 'cumulative_coal_co2',
 'cumulative_flaring_co2',
 'cumulative_gas_co2',
 'cumulative_luc_co2',
 'cumulative_oil_co2',
 'cumulative_other_co2',
 'energy_per_capita',
 'energy_per_gdp',
 'flaring_co2',
 'flaring_co2_per_capita',
 'gas_co2',
 'gas_co2_per_capita',
 'ghg_excluding_lucf_per_capita',
 'ghg_per_capita',
 'land_use_change_co2',
 'land_use_change_co2_per_capita',
 'methane',
 'methane_per_capita',
 

### Selecting 5 columns for further analysis: population, country, year, co2, and co2_per_capita for my analysis.

In [5]:
df.shape

(50411, 79)

### Let's see the structure

In [6]:
df.head(10)

,country,year,iso_code,population,gdp,cement_co2,cement_co2_per_capita,co2,co2_growth_abs,co2_growth_prct,...,share_global_other_co2,share_of_temperature_change_from_ghg,temperature_change_from_ch4,temperature_change_from_co2,temperature_change_from_ghg,temperature_change_from_n2o,total_ghg,total_ghg_excluding_lucf,trade_co2,trade_co2_share
0,Afghanistan,1750,AFG,2802560.0,NaN,0.0,0.0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Afghanistan,1751,AFG,NaN,NaN,0.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,Afghanistan,1752,AFG,NaN,NaN,0.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Afghanistan,1753,AFG,NaN,NaN,0.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,Afghanistan,1754,AFG,NaN,NaN,0.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
5,Afghanistan,1755,AFG,NaN,NaN,0.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,Afghanistan,1756,AFG,NaN,NaN,0.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7,Afghanistan,1757,AFG,NaN,NaN,0.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8,Afghanistan,1758,AFG,NaN,NaN,0.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
9,Afghanistan,1759,AFG,NaN,NaN,0.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### In data there are missing values. Veifying whether they are also in the selected columns

In [7]:
selected_columns = [
    "country",
    "year",
    "population",
    "co2",
    "co2_per_capita"
]

df[selected_columns].isna().sum()

country               0
year                  0
population         9244
co2               21027
co2_per_capita    23902
dtype: int64

### Some missing values are present in the selected columns.
### The next step is to verify whether they affect Poland or the EU27 observation

In [8]:
min_year = df["year"].min()
max_year = df["year"].max()

print(f"Minimum year is {min_year}, and maximum is {max_year}")

Minimum year is 1750, and maximum is 2024


### The latest data is 2024, and earliest 1750. Checking data types

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50411 entries, 0 to 50410
Data columns (total 79 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   country                                    50411 non-null  object 
 1   year                                       50411 non-null  int64  
 2   iso_code                                   42480 non-null  object 
 3   population                                 41167 non-null  float64
 4   gdp                                        15251 non-null  float64
 5   cement_co2                                 29173 non-null  float64
 6   cement_co2_per_capita                      25648 non-null  float64
 7   co2                                        29384 non-null  float64
 8   co2_growth_abs                             27216 non-null  float64
 9   co2_growth_prct                            26239 non-null  float64
 10  co2_including_luc     

### Checking unique country values to see whether we look for "EU", or "European Union" etc.

In [10]:
sorted(df["country"].unique())

['Afghanistan',
 'Africa',
 'Africa (GCP)',
 'Albania',
 'Algeria',
 'Andorra',
 'Angola',
 'Anguilla',
 'Antarctica',
 'Antigua and Barbuda',
 'Argentina',
 'Armenia',
 'Aruba',
 'Asia',
 'Asia (GCP)',
 'Asia (excl. China and India)',
 'Australia',
 'Austria',
 'Azerbaijan',
 'Bahamas',
 'Bahrain',
 'Bangladesh',
 'Barbados',
 'Belarus',
 'Belgium',
 'Belize',
 'Benin',
 'Bermuda',
 'Bhutan',
 'Bolivia',
 'Bonaire Sint Eustatius and Saba',
 'Bosnia and Herzegovina',
 'Botswana',
 'Brazil',
 'British Virgin Islands',
 'Brunei',
 'Bulgaria',
 'Burkina Faso',
 'Burundi',
 'Cambodia',
 'Cameroon',
 'Canada',
 'Cape Verde',
 'Central African Republic',
 'Central America (GCP)',
 'Chad',
 'Chile',
 'China',
 'Christmas Island',
 'Colombia',
 'Comoros',
 'Congo',
 'Cook Islands',
 'Costa Rica',
 "Cote d'Ivoire",
 'Croatia',
 'Cuba',
 'Curacao',
 'Cyprus',
 'Czechia',
 'Democratic Republic of Congo',
 'Denmark',
 'Djibouti',
 'Dominica',
 'Dominican Republic',
 'East Timor',
 'Ecuador',
 'Egy

### Data includes European Union with 27, or 28 countries. EU 27 countries will be analyzed (without Great Britain).

In [11]:
country_filter = df["country"].isin(["Poland", "European Union (27)"])
year_filter = df["year"] >= 1990

filtered_df = df[country_filter & year_filter]

### Filtering data for desired columns: population, year, country, co2, and co2_per_capita

In [14]:
filtered_df = filtered_df[selected_columns]

### Check to see whether there are still some missing values

In [15]:
filtered_df.isna().sum()

country           0
year              0
population        0
co2               0
co2_per_capita    0
dtype: int64

In [16]:
filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 70 entries, 16233 to 37610
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   country         70 non-null     object 
 1   year            70 non-null     int64  
 2   population      70 non-null     float64
 3   co2             70 non-null     float64
 4   co2_per_capita  70 non-null     float64
dtypes: float64(3), int64(1), object(1)
memory usage: 3.3+ KB


### The population column is stored as float64 because the original dataset contains missing values for some countries and years. The filtered dataset contains no missing values, so the column could be safely converted to int64 if required.

In [17]:
filtered_df.head(10)

,country,year,population,co2,co2_per_capita
16233,European Union (27),1990,419997519.0,3870.253,9.215
16234,European Union (27),1991,421419173.0,3804.530,9.028
16235,European Union (27),1992,422863649.0,3677.736,8.697
16236,European Union (27),1993,424104740.0,3608.154,8.508
16237,European Union (27),1994,425189040.0,3589.454,8.442
16238,European Union (27),1995,425847857.0,3635.878,8.538
16239,European Union (27),1996,426181943.0,3720.710,8.730
16240,European Union (27),1997,426651685.0,3654.451,8.565
16241,European Union (27),1998,426655455.0,3644.649,8.542
16242,European Union (27),1999,426830677.0,3589.161,8.409


### There could be a possibility to convert 64 bit data types into something smaller to save space if required. I also noticed old index is kept. Resseting the old index.


In [18]:
filtered_df = filtered_df.reset_index(drop=True)

### Additional checks to see whether data is correct (amount of rows, countries picked, no duplicates)

In [19]:
filtered_df.shape

(70, 5)

In [20]:
filtered_df["country"].unique()

array(['European Union (27)', 'Poland'], dtype=object)

In [21]:
filtered_df["country"].value_counts()

country
European Union (27)    35
Poland                 35
Name: count, dtype: int64

In [22]:
filtered_df.duplicated().sum()

0

In [23]:
filtered_df.groupby("country")["year"].count()

country
European Union (27)    35
Poland                 35
Name: year, dtype: int64

## Data quality summary

- Filtered to Poland and European Union (27)
- Period: 1990–2024
- Dataset shape: 70 rows × 5 columns
- No missing values detected
- No duplicate rows detected
- Each entity contains a complete 35-year time series

## Saving filtered data into processed folder

In [24]:
filtered_df.to_csv(
    "../data/processed/co2_poland_eu.csv",
    index = False
)

### Check if the file was saved correctly

In [29]:
pd.read_csv("../data/processed/co2_poland_eu.csv").head(10)

,country,year,population,co2,co2_per_capita
0,European Union (27),1990,419997519.0,3870.253,9.215
1,European Union (27),1991,421419173.0,3804.530,9.028
2,European Union (27),1992,422863649.0,3677.736,8.697
3,European Union (27),1993,424104740.0,3608.154,8.508
4,European Union (27),1994,425189040.0,3589.454,8.442
5,European Union (27),1995,425847857.0,3635.878,8.538
6,European Union (27),1996,426181943.0,3720.710,8.730
7,European Union (27),1997,426651685.0,3654.451,8.565
8,European Union (27),1998,426655455.0,3644.649,8.542
9,European Union (27),1999,426830677.0,3589.161,8.409
